File: go/linked_list/0021_merge_two_sorted_lists/merge_two_sorted_lists.go
package mergetwosortedlists

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	pointer := dummy

	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			pointer.Next = list1
			list1 = list1.Next
		} else {
			pointer.Next = list2
			list2 = list2.Next
		}
		pointer = pointer.Next
	}

	if list1 != nil {
		pointer.Next = list1
	} else {
		pointer.Next = list2
	}

	return dummy.Next
}

File: c/linked_list/0021_merge_two_sorted_lists/merge_two_sorted_lists.c
#include <stdlib.h>
struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *mergeTwoLists(struct ListNode *list1, struct ListNode *list2)
{
    struct ListNode dummy = {
        0,
        NULL};

    struct ListNode *pointer = &dummy;

    while (list1 != NULL && list2 != NULL)
    {
        if (list1->val < list2->val)
        {
            pointer->next = list1;
            list1 = list1->next;
        }
        else
        {
            pointer->next = list2;
            list2 = list2->next;
        }
        pointer = pointer->next;
    }

    if (list1 != NULL)
    {
        pointer->next = list1;
    }
    else
    {
        pointer->next = list2;
    }

    return dummy.next;
}